lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,217 @@
+    # @param {String} token: Token to check.
+    # @return {Boolean} False if token is in list, true otherwise.
+    check: (token) =>
+      correctionsList = @getJson()
+      # New JSON structure makes it possible to search for token
+      # just like a dictionary key.
+      return !correctionsList[token]
 
+    # Return corrections list for token if exists.
+    # @param {String} token: token to search in corrections list from server.
+    # @return {Array}: list of corrections.
+    getCorrections: (token) ->
+      return if not @check(token) then @getJson()[token]
 
-    exports.SpellChecker = SpellChecker
+
+  # Sets up spellchecker popup and implements some routines
+  # to work on current in the editor.
+  setupSpellCheckerPopup = (editor) ->
+    PopupManager.init(editor)
+
+    # Check if current word is in corrections list.
+    # Call PopupManager.show if it's present, do nothing otherwise.
+    newPopup = ->
+      word = extractWord(editor)
+      spellChecker = new SpellChecker()
+      correctionsItem = spellChecker.getCorrections(word)
+      if correctionsItem
+        PopupManager.show(convertCorrectionList(correctionsItem))
+      return
+
+    # Bind newPopup to Alt-Enter editor shortcut.
+    editor.commands.addCommand({
+      name: "newPopup",
+      bindKey: "Alt-Enter",
+      exec: newPopup
+    })
     return
-)
\ No newline at end of file
+
+
+  # Get the word under the cursor.
+  # @param {Editor} editor: editor object.
+  # @return {String}: the current word.
+  extractWord = (editor) ->
+    session = editor.getSession()
+    wordRange = getCurrentWordRange(editor)
+    return session.getTextRange(wordRange)
+
+  # Returns Range object that describes the current word position.
+  # @param {Editor} editor: editor object.
+  # @return {Range}: current word range.
+  getCurrentWordRange = (editor) ->
+    session = editor.getSession()
+    row = editor.getCursorPosition().row
+    col = editor.getCursorPosition().column
+    return session.getWordRange(row, col)
+
+  # Convert an array of string to popup-eligible structure.
+  # @param {Array} corrections: array of strings with substitution options.
+  # @return {Array}: array of JSONs, actually.
+  convertCorrectionList = (corrections) ->
+    return ({caption: item, value: item} for item in corrections)
+
+
+  # Object contains init, show and detach functions for popup
+  # and some routines for inner use.
+  PopupManager =
+    # Bindings for editor keys with PopupManager functions.
+    commands:
+      "Esc": ->
+        PopupManager.detach()
+      "Up": ->
+        PopupManager.goTo("up")
+      "Down": ->
+        PopupManager.goTo("down")
+      "Return": ->
+        PopupManager.insertCorrection()
+
+    # Initial setup for popup.
+    # @param {Editor} editor: editor object.
+    init: (editor) ->
+      @editor = editor
+      @activated = false
+
+      HashHandler = require("ace/keyboard/hash_handler").HashHandler
```

Comment:
I think it won't since I tried and got `Failed to execute 'importScripts' on 'WorkerGlobalScope'` as always when I try to require these modules at the file's global scope.